cli/tests/stake.rs

```diff
@@ -2320,3 +2320,757 @@ fn test_stake_minimum_delegation() {
     let result = process_command(&config);
     assert!(matches!(result, Ok(..)));
 }
+
+#[test]
+fn test_stake_redelegation_then_deactivation_withdraw_not_permitted() {
```

Comment:
I am not sure that I understand your comments. 

1. Inside cli/tests, it uses `TestValidator` class. Is that not the `SolanaTestValidator`?
2. The test scenario for this bug is very complex. It involves multiple stake accounts and votes accounts, and a sequence of very specific actions, i.e. delegate, redelegate, merge, deactivate, withdraw etc. It also requires stake_history mutation. I am not sure how this can be modeled on just bank itself. I think bank tests are more suitable to test individual instruction. Do you know any examples on bank test to model stake deactivation, activation and withdraw?
3. It seems like cli/stake.rs is where we are testing all other existing scenarios of stake manipulation, such as redelegation, withdraw etc. Therefore, I feel it is natural to extend those tests to cover the new behavior.

Reply:
i'm saying that the cli crate is not a test framework, that these tests belong elsewhere and implemented without the cli details

Reply:
To me, the cli/tests looks more or less like a test framework. It has been used to test stake, vote, transfer, and other programs.  And it has very useful `test_util.rs`, which make it very convenient to model complex test scenario there. 

I don't know if we should "reinvent the wheel" to write another test framework to model the lifetime of a test validator. Or maybe I am missing something about the other test framework?
 
Do you know where the other test framework/utility is, which could be used to better model these tests?

Reply:
> To me, the cli/tests looks more or less like a test framework.

this is a function of those tests existing before `TestValidator` and us for some reason being very paranoid about cli behaving as expected. most, if not all of them should really be modernized at some point to avoid leading folks astray

> I don't know if we should "reinvent the wheel" to write another test framework to model the lifetime of a test validator.

we don't need to, that's literally what `TestValidator` was for.  the tests will be similar but won't a) live under the totally unrelated cli crate and b) have all of the superfluous cli machination in the way and confounding tests. you'll grab an `RpcClient` handle from the `TestValidator` instance, build the txes up and `rpc_client.send_and_confirm_transaction()` your way to victory. these should probably live under `programs/stake`, either a new `tests` directory or if one of existing source files seems appropriate

it should be a red flag when tests are being added to a crate that was otherwise unaffected by the change set that something isn't right organizationally

Reply:
ok. i will take a look.

Reply:
I pushed a commit to move the tests from cli to `program/stake`. 

But the ci sanity check is not happy about it. :(

Adding solana-test-validator to solana-stake-program results a circular dependency, i.e. 
solana-test-validator --> solana-runtime --> solana-stake-program --> solana-test-validator!

@t-nelson

Reply:
ugh... runtime doesn't need `solana-stake-program` all of those symbols should be available from sdk.  it must be using the re-exports instead of having been properly migrated.

i'll take a look tomorrow

Reply:
started in https://github.com/solana-labs/solana/pull/32635. need to figure out the best home for the remaining symbols

Reply:
This is a common issue, and rather than leaning into the sdk more, we should resolve it using the same technique as the address-lookup-table program, which has a separate crate for tests.

In my opinion, the program crates like `solana-stake-program` should function like normal program crates, like `spl-token`. When you want those types and instructions, you pull in that crate, rather than relying on `solana-program` to have everything in it.

I know that the separate crate for tests is ugly, but it avoids bloating the sdk and resolves the circular dependencies.

Reply:
It seems that moving the stake tests out of `cli` is not as straightforward as it should be.

How about keeping these 3 tests in `cli` for this PR, and moving all stake tests out of `cli` in a separated follow-up PR?

Reply:
> which has a separate crate for tests

oh thank god. i missed that this wasn't implementing the tests directory as a standalone crate. just realized i was about to get screwed by the instruction processor being globbed into `BUILTINS`. this is The Way :tm: 

> the program crates like solana-stake-program should function like normal program crates

generally agree, but there's a bunch of crap in this program crate that makes zero sense. even have comments about being external helpers with uses listed that are... not the program itself!

Reply:
> It seems that moving the stake tests out of `cli` is not as straightforward as it looks.
> 
> How about keeping these 3 tests in `cli` for this PR, and moving all stake tests out of `cli` in a separated follow-up PR after we fixes the circular dependencies issue?

take a crack at making the tests dir a crate as jon proposes first. doing the easy thing our of impatience is how the code got into the shape it is today. that followup pr rarely happens

Reply:
ok. I will create a new test crate `solana-stake-program-tests` for those tests.

Reply:
> generally agree, but there's a bunch of crap in this program crate that makes zero sense. even have comments about being external helpers with uses listed that are... not the program itself!

That's totally fair... the native program crates are probably the most problematic with dependencies. We could eventually make it a policy that *all* native stuff goes into the sdk, but I would prefer to thin out the sdk whenever possible.

Note: I'm the one who moved a lot of the stake types into the sdk in the first place while developing stake pools, because I didn't think to use `target_os` guards in `solana-stake-program`, so I bear a lot of the responsibility there :sweat:

Reply:
Done!